File: medpass/src/types/calendar.ts
export interface RecurrenceRule {
  frequency: string; 
  interval?: number;
  days_of_week?: number[]; 
  end_date?: string; 
  count?: number;
}

export interface CalendarEvent {
  id: string;
  student_id?: number; 
  title: string;
  description?: string;
  start: string;
  end: string;
  allDay: boolean; 
  type: string; 
  recurrence?: RecurrenceRule | null; 
  location?: string;
  color?: string;
  priority?: number;
  topicName?: string;
  completed?: boolean;
  createdAt?: string; 
  updatedAt?: string; 

  extendedProps?: {
    [key: string]: any;
    type: string;
    description?: string;
    location?: string;
    priority?: number;
    topicName?: string;
    completed?: boolean;
  };
}

export interface StudySession extends CalendarEvent {
  topicName: string;
  completed: boolean;
  plan_id?: string;
}

export interface EventCreatePayload {
  title: string;
  description?: string;
  start_time: string; 
  end_time: string; 
  all_day?: boolean;
  event_type: string;
  recurrence?: RecurrenceRule | null;
  location?: string;
  color?: string;
  priority?: number;
  topic_name?: string; 
  completed?: boolean;
}

export interface EventUpdatePayload extends Partial<EventCreatePayload> {}

export interface WeaknessStrength {
  subject: string;
  performance: number; 
  is_weakness?: boolean; 
  unit_type?: string; 
  performance_score?: number; 
}

export interface MLPrediction {
  prediction: number; 
  probability: number;
  prediction_text: string;
  confidence_score: number;
}

export interface RiskAssessment {
  risk_score: number;
  risk_level: 'Low' | 'Medium' | 'High';
  strengths: WeaknessStrength[];
  weaknesses: WeaknessStrength[];
  ml_prediction: MLPrediction;
  details?: {
    student_name?: string;
    total_exams?: number;
    passed_exams?: number;
    total_grades?: number;
    ml_model_accuracy?: number;
  };
}

export interface StudyPlanRequestPayload {
  exam_date: string; 
  weaknesses: WeaknessStrength[];
  strengths: WeaknessStrength[];
  events: CalendarEvent[];
  study_hours_per_day?: number;
  focus_areas?: string[];
  additional_notes?: string; 
}

export interface StudyPlanEventResponse {
  id: string;
  title: string;
  description?: string;
  start: string; 
  end: string; 
  allDay: boolean;
  type: string; 
  topicName?: string;
  completed?: boolean;
}

export interface StudyPlanResponseData {
  id: string;
  title: string;
  description?: string;
  startDate: string;
  endDate: string;
  examDate: string; 
  events: StudyPlanEventResponse[];
  createdAt: string;
}

export interface StudyPlanSummary {
  total_study_hours: number;
  topics_count: number;
  focus_areas: { [key: string]: number };
  weekly_breakdown: { [key: string]: number };
}

export interface StudyPlanGenerationResponse {
  plan: StudyPlanResponseData;
  summary: StudyPlanSummary;
}

// StudyPlanAnalytics displays progress over time
export interface StudyPlanProgress {
  total_sessions: number;
  completed_sessions: number;
  completion_percentage: number;
  hours_studied: number;
  hours_remaining: number;
  progress_by_topic: { [key: string]: number }; // percentage completed by topic
}

// Type for API interaction functions
export interface CalendarApi {
  getEvents: () => Promise<CalendarEvent[]>;
  createEvent: (payload: EventCreatePayload) => Promise<CalendarEvent>;
  updateEvent: (eventId: string, payload: EventUpdatePayload) => Promise<CalendarEvent>;
  deleteEvent: (eventId: string) => Promise<{ message: string }>;
  getRiskAssessment: () => Promise<RiskAssessment>; 
  generateStudyPlan: (payload: StudyPlanRequestPayload) => Promise<StudyPlanGenerationResponse>;
  markEventCompleted: (eventId: string, completed: boolean) => Promise<CalendarEvent>;
  exportPlanAsPdf: (planId: string) => Promise<Blob>;
  getStudyPlanProgress: (planId: string) => Promise<StudyPlanProgress>;
}

// Type for FullCalendar event mapping
export interface FullCalendarEvent {
  id: string;
  title: string;
  start: string;
  end: string;
  allDay: boolean;
  color?: string;
  extendedProps: {
    [key: string]: any;
    type: string;
    description?: string;
    location?: string;
    priority?: number;
    topicName?: string;
    completed?: boolean;
  };
}

File: backend/app/schemas/calendar.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any, Union
from datetime import datetime
import uuid

class RecurrenceRule(BaseModel):
    frequency: str
    interval: Optional[int] = None
    days_of_week: Optional[List[int]] = None
    end_date: Optional[datetime] = None
    count: Optional[int] = None

class EventBase(BaseModel):
    title: str
    description: Optional[str] = None
    start_time: datetime
    end_time: datetime
    all_day: Optional[bool] = False
    event_type: str
    recurrence: Optional[RecurrenceRule] = None
    location: Optional[str] = None
    color: Optional[str] = None
    priority: Optional[int] = None

class EventCreate(EventBase):
    pass

class EventUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    all_day: Optional[bool] = None
    event_type: Optional[str] = None
    recurrence: Optional[RecurrenceRule] = None
    location: Optional[str] = None
    color: Optional[str] = None
    priority: Optional[int] = None

class Event(BaseModel):
    id: str
    title: str
    description: Optional[str] = None
    start: str
    end: str
    all_day: bool
    type: str
    recurrence: Optional[Dict[str, Any]] = None
    location: Optional[str] = None
    color: Optional[str] = None
    priority: Optional[int] = None

    class Config:
        from_attributes = True

class StudyPlanEventBase(BaseModel):
    """Schema for study plan event fields"""
    topic_id: Optional[str] = None
    topic_name: Optional[str] = None
    difficulty: Optional[int] = None
    importance: Optional[int] = None
    completed: Optional[bool] = False

class StudyPlanEventCreate(StudyPlanEventBase, EventCreate):
    pass

class StudyPlanEventUpdate(StudyPlanEventBase, EventUpdate):
    pass

class StudyPlanEventResponse(BaseModel):
    """Schema for study plan event response"""
    id: str
    title: str
    description: Optional[str] = None
    start: str 
    end: str 
    allDay: bool
    type: str 
    topicName: Optional[str] = None
    completed: Optional[bool] = False

class StudyPlanBase(BaseModel):
    """Base schema for study plan"""
    title: str
    description: Optional[str] = None
    start_date: datetime
    end_date: datetime
    exam_date: datetime

class StudyPlanCreate(StudyPlanBase):
    """Schema for creating study plan"""
    events: List[StudyPlanEventCreate]

class StudyPlanUpdate(BaseModel):
    """Schema for updating study plan"""
    title: Optional[str] = None
    description: Optional[str] = None
    start_date: Optional[datetime] = None
    end_date: Optional[datetime] = None
    exam_date: Optional[datetime] = None

class StudyPlanInDB(StudyPlanBase):
    """Schema for study plan in database"""
    plan_id: str
    student_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

class StudyPlanResponseData(BaseModel):
    """Schema for study plan response data"""
    id: str
    title: str
    description: Optional[str] = None
    startDate: str
    endDate: str
    examDate: str 
    events: List[StudyPlanEventResponse]
    createdAt: str

class WeaknessStrength(BaseModel):
    """Schema for student strength/weakness"""
    subject: str
    performance: float
    is_weakness: Optional[bool] = None
    unit_type: Optional[str] = None
    performance_score: Optional[float] = None

class StudyPlanRequest(BaseModel):
    """Schema for study plan generation request"""
    exam_date: datetime
    weaknesses: List[WeaknessStrength]
    strengths: List[WeaknessStrength]
    events: List[Dict[str, Any]]
    study_hours_per_day: Optional[int] = 4
    focus_areas: Optional[List[str]] = None
    additional_notes: Optional[str] = None

class StudyPlanSummary(BaseModel):
    """Schema for study plan summary"""
    total_study_hours: int
    topics_count: int
    focus_areas: Dict[str, int]
    weekly_breakdown: Dict[str, int]

class StudyPlanResponse(BaseModel):
    """Schema for complete study plan response"""
    plan: StudyPlanResponseData
    summary: StudyPlanSummary

def generate_uuid() -> str:
    """Generate a UUID string"""
    return str(uuid.uuid4())